### ISSUE TITLE
Deserialization of generics is broken on 2.7.3 if custom abstract type resolver is set

### ISSUE THREAD

**Issue by j-j-marvin** [FasterXML/jackson-databind#1173]
Hi there,

After migration from 2.6.1 to 2.7.3 I've found that all deserialization of generic types are broken in cases when custom type resolver is specified. I've created a sample to reproduce the issue.

On 2.6.1 the output is:
`class com.dummy.test.DeserializationTest$B`

Where as on 2.7.3 it has changed to:
`class com.dummy.test.DeserializationTest$MyMap`

If I comment `mapper.registerModule(basicModule);` on 2.7.3 then deserialization also works as expected (the same way as on 2.6.1).

```java
package com.dummy.test;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleAbstractTypeResolver;
import com.fasterxml.jackson.databind.module.SimpleModule;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;


public class DeserializationTest {

    public static void main(String[] args) throws IOException {
        ObjectMapper mapper = new ObjectMapper();
        SimpleAbstractTypeResolver resolver = new SimpleAbstractTypeResolver();
        resolver.addMapping(Map.class, MyMap.class);

        SimpleModule basicModule = new SimpleModule();
        basicModule.setAbstractTypes(resolver);
        mapper.registerModule(basicModule);
        String value = "{\"z\": {\"zz\": {\"a\": 42}}}";
        A a = mapper.readValue(value, A.class);

        Class deserializedClass = a.getMap().get("zz").getClass();
        System.out.println(deserializedClass);
    }

    public static class A {
        private Map map;

        @JsonCreator
        public A(@JsonProperty("z") Map<String, B> map) {
            this.map = map;
        }

        public Map getMap() {
            return map;
        }
    }

    public static class B {
        private int a;

        public B(@JsonProperty("a") int a) {
            this.a = a;
        }
    }

    public static class MyMap extends HashMap {
    }
}
```

**Comment by cowtowncoder** [comment]
Thank you for reporting this. It sounds plausible, as type resolution was fully rewritten between 2.6 and 2.7.

**Comment by cowtowncoder** [comment]
Ok, actually the problem is with the code -- although things appeared to work in 2.6, it was by accident more than design. The problem is this:

```java
public static class MyMap extends HashMap { }
```

which leaves out type parameters. As such, it must resolve to `MyMap extends HashMap<Object,Object> { }`, and parameterization can not be promoted from base type, despite there being `Map<String,B>`.

So, the fix needs to be to include nominal type parameters, a la:

```java
public static class MyMap<K,V> extends HashMap<K,V> { }
```

with which test passes.

**Comment by j-j-marvin** [comment]
> As such, it must resolve to `MyMap extends HashMap<Object,Object> { }`, and parameterization can not be promoted from base type, despite there being `Map<String,B>`.
Yep, I see, however, types are available from target constructor (and I assume in 2.6.* taken from there). Also I don't see any contradiction in such assignment, nevertheless I understand that without  full codebase this sample looks a little bit strange.
One more thing to be more precise, in our case `MyMap extends HashMap<String, Object>` and cannot be converted to `MyMap<V> extends HashMap<String, V>`, since it is ok in some places to have values of different type in such map.
So the only way I see on 2.7.3 is to use different Mapper setup when extracting hierarchy of MyMap and normal classes, which is generally fine. However, still not clear why type cannot be extracted from target argument and looks like loosing a feature.

**Comment by cowtowncoder** [comment]
@j-j-marvin From conceptual perspective your class hierarchy can not produce outcome you want -- if `MyMap` has parameterization of `Map<?,?>`, it is not possible to refine from `Map<String,B>` into `MyMap` at all: type parameters are not compatible.

The reason things worked earlier was because type handling system took various short-cuts and assumed that all `MapType` types, subtypes, always have 2 type parameters corresponding to Key and Value. This is not true, for, say, type `class StringKeyMap<V> extends HashMap<String,V>' where types are partially bound. So the conceptually sound method of resolving type parameters up to subtype and then down to type caller is interested in (in this case `Map`, but that is not something that is known at resolution time) would detect a problem that a simpler system does not.

So why more complex system? Because there are many other cases that could not be solved by it, especially regarding type aliasing (renaming of type variables) and some cyclic types.

So long story short: yes, it is unfortunate your usage has changed and something that did work does not. If I felt this can be fixed in solid way, I would do so. I will actually keep this in mind, still, it is plausible I might have overlooked something.
But in the meantime I would suggest adding type variables as it seems to me that they really should exist.

**Comment by j-j-marvin** [comment]
@cowtowncoder thanks for details answer. Meanwhile, I've split my usage of ObjectMapper in two streams one for loading normal hierarchies of Map<K, V> and the other of MyMap only. Now I think it looks better from design point of view as well.

However, I'd suggest to specify this change in 2.7 release notes.
